lib/ui/window.dart

```diff
@@ -143,14 +161,234 @@ class Locale {
   /// ```
   ///
   /// The primary language subtag must not be null. The region subtag is
-  /// optional.
-  ///
-  /// The values are _case sensitive_, and should match the case of the relevant
-  /// subtags in the [IANA Language Subtag
-  /// Registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry).
-  /// Typically this means the primary language subtag should be lowercase and
-  /// the region subtag should be uppercase.
-  const Locale(this._languageCode, [ this._countryCode ]) : assert(_languageCode != null);
+  /// optional. The values are _case sensitive_, and must match the values
+  /// listed in
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/language.xml and
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/region.xml.
+  ///
+  /// This method only produces standards-compliant instances if valid language
+  /// and country codes are provided. Deprecated subtags will be replaced, but
+  /// incorrectly cased strings are not corrected.
+  const Locale(
+    this._languageCode, [
+    this._countryCode,
+  ]) : assert(_languageCode != null),
+       this._scriptCode = null,
+       this._variants = null,
+       this._extensions = null;
+
+  /// Creates a new Locale object with the specified parts.
+  ///
+  /// This is for internal use only. All fields must already be normalized, must
+  /// already be canonicalized. This method does not modify parameters in any
+  /// way or do any syntax checking.
+  ///
+  /// * language, script and region must be in canonical form.
+  /// * iterating over variants must provide variants in alphabetical order.
+  /// * The extensions map must contain only valid key/value pairs. "u" and "t"
+  ///   keys must be present, with an empty string as value, if there are any
+  ///   subtags for those singletons.
+  Locale._internal(
+    String language, {
+    String script,
+    String region,
+    collection.SplayTreeSet<String> variants,
+    collection.SplayTreeMap<String, String> extensions,
+  }) : assert(language != null),
+       assert(language.length >= 2),
+       assert(language.length <= 8),
+       assert(language.length != 4),
```

Comment:
are you sure the length of language codes will _never_ change? This seems like a surprising thing to enforce so strictly.

Reply:
Yes. We apparently have <9000 valid Language codes presently[1], they are all 2 to 3 characters. The 5-8 character possibilities are for future expansions, several orders of magnitude for future expansion.
[1] http://unicode.org/repos/cldr/tags/latest/common/validity/language.xml

Reply:
I don't really understand why we would want to enforce it in this API. Why this but not, say, that the codepoints aren't unpaired UTF-16 surrogates?

It seems like if we want to do validation we should have a separate validation step (ideally not in dart:i18n but in some package that you can link in if you need it).

Reply:
(note in particular that these asserts are debug-time only, so it's not like we're even actually checking anything in release builds)

Reply:
This _internal constructor is currently only used by parse(). Parse rejects, by throwing an exception, anything that doesn't match regular expressions. Thus parse() is doing the syntax checking and these asserts are intended as just a debug-time double-check, rather than as programmatic validation.

Reply:
I dropped asserts in https://github.com/flutter/engine/pull/6518 - considering this to generally be "handled/decided".

(Speculation: migration to dart:i18n may bring up new questions w.r.t. JavaScript interop.)